File: app/src/main/java/com/beibei/init/common/networkJava/HttpUtil.java
package com.beibei.init.common.networkJava;

import com.beibei.init.common.constant.Constant;
import com.beibei.init.common.network.model.ResultBean;
import com.beibei.init.common.utils.LogUtils;
import com.beibei.init.common.utils.SignUtil;

import java.util.HashMap;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

;

public class HttpUtil {
    private static volatile HttpUtil httpUtil = null;

    private HttpUtil() {
    }

    public static HttpUtil getInstance() {
        if (httpUtil == null) {
            synchronized (HttpUtil.class) {
                if (httpUtil == null) {
                    httpUtil = new HttpUtil();
                }
            }
        }
        return httpUtil;
    }

    public <T> void request(Observable<ResultBean<T>> observable, final OnResultListener<T> onResultListener) {
        // 步骤7：发送网络请求
        observable.subscribeOn(Schedulers.io())               // 在IO线程进行网络请求
                .observeOn(AndroidSchedulers.mainThread())  // 回到主线程 处理请求结果
                .subscribe(new Observer<ResultBean<T>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(ResultBean<T> result) {
                        onResultListener.onSuccess(result.getData());
                    }

                    @Override
                    public void onError(Throwable e) {
                        onResultListener.onError(e);
                    }

                    @Override
                    public void onComplete() {
//                        Log.d(TAG, "请求成功");
                    }
                });
    }

    public interface OnResultListener<T> {
        void onSuccess(T result);

        void onError(Throwable error);

        void onMessage();
    }

    public HashMap<String, String> dataDealWith(String data) {
        LogUtils.e("------------OkHttp---------data-------" + data);
        String partner = Constant.PARTNER;
        String sign = SignUtil.dataDealWith(data);
        HashMap<String, String> map = new HashMap<>();
        map.put("data", data);
        map.put("partner", partner);
        map.put("sign", sign);
        return map;
    }
}

File: app/src/main/java/com/beibei/init/common/network/HttpUtil.kt
package com.beibei.init.common.newNetwork

import com.beibei.init.App
import com.beibei.init.common.constant.Constant
import com.beibei.init.common.network.model.ResultBean
import com.beibei.init.common.newNetwork.exception.ErrorType
import com.beibei.init.common.newNetwork.exception.ExceptionEngine
import com.beibei.init.common.utils.LogUtils
import com.beibei.init.common.utils.SignUtil
import com.dashen.demeter.common.utils.NetUtils
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File

/**
 * Created by beibei on 2018/3/15.
 */
object HttpUtil {
    fun <T> request(observable: Observable<ResultBean<T>>, resultListener: OnResultListener<T>) {
        setSubscriber(observable, object : Observer<ResultBean<T>> {
            override fun onComplete() {
                LogUtils.e("-----onCompleted-----" + "读取完成")
            }

            override fun onSubscribe(d: Disposable) {
            }

            override fun onNext(bean: ResultBean<T>) {
                resultListener.onSuccess(bean.data)
                resultListener.onMessage(bean.errcode, bean.message ?: "----exception---后台没给传message--")
            }

            override fun onError(error: Throwable) {
                val apiException = ExceptionEngine.handleException(error)
                LogUtils.e("---------exception--------errorCode:" + apiException.errorCode + ",message:" + apiException.message)
                var msg: String
                msg = when (apiException.errorCode) {
                    ErrorType.HTTP_ERROR -> "服务器连接异常，请稍后重试"//协议错误
                    ErrorType.NETWORD_ERROR -> {//网络错误
                        if (NetUtils.getNetWorkState(App.context) == NetUtils.NETWORK_NONE) { //网络连接
                            "网络不给力，请检查网络连接"
                        } else {  //服务器异常
                            "服务器连接异常，请稍后重试"
                        }
                    }
                    ErrorType.PARSE_ERROR -> "数据加载错误，请稍后再试"//解析错误
                    ErrorType.UNKNOWN -> "数据加载错误，请稍后再试"//未知错误
                    else -> apiException.message!!//服务器返回errocode不为0
                }
                resultListener!!.onError(apiException, msg)
                resultListener.onMessage(apiException.errorCode, msg)
            }
        })
    }

    //文件转为可上传的文件流体
    fun covertFile2Body(fileParamName: String, file: File): MultipartBody.Part {
        //构建requestbody
        val requestFile = RequestBody.create(MediaType.parse("application/otcet-stream"), file)
        //将resquestbody封装为MultipartBody.Part对象
        return MultipartBody.Part.createFormData(fileParamName, file.name, requestFile)
    }

    //订阅事件
    private fun <T> setSubscriber(observable: Observable<ResultBean<T>>, observer: Observer<ResultBean<T>>) {
        observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(observer)
    }


//    去掉
    fun dealData(data: String): HashMap<String, String> {
        LogUtils.e("--OkHttp---data-" + data)
        val partner = Constant.PARTNER
        val sign = SignUtil.dataDealWith(data)

        val hashMap = HashMap<String, String>()
        hashMap.put("data", data)
        hashMap.put("partner", partner)
        hashMap.put("sign", sign)
        return hashMap
    }

    //接口回调
    interface OnResultListener<T> {
        /**
         * 成功回调方法
         * @param t
         */
        fun onSuccess(t: T?)

        /**
         * 失败回调方法
         * @param error
         * *
         * @param msg
         */
        fun onError(error: Throwable, msg: String)

        /**
         * 输出msg(肯定会执行)
         */
        fun onMessage(errorCode: Int, msg: String)
    }
}